### PR TITLE
KEP-572: Implement VolumeClaimPolicies API for Stateful JobSet

### PR DESCRIPTION
**Pull request by andreyvelich**
#### What type of PR is this?

/kind api-change

#### What this PR does / why we need it:

This PR implemented VolumeClaimPolicies API to manage volumes for JobSet. Working example:
```yaml
apiVersion: jobset.x-k8s.io/v1alpha2
kind: JobSet
metadata:
  name: volume-claim-trainjob
spec:
  volumeClaimPolicies:
    - templates:
        - metadata:
            name: initializer
          spec:
            accessModes: ["ReadWriteOnce"]
            resources:
              requests:
                storage: 1Gi
    - templates:
        - metadata:
            name: checkpoints
          spec:
            accessModes: ["ReadWriteOnce"]
            resources:
              requests:
                storage: 1Gi
      retentionPolicy:
        whenDeleted: Retain
  replicatedJobs:
    - name: initializer
      template:
        spec:
          template:
            spec:
              containers:
                - name: initializer
                  image: busybox
                  command:
                    - /bin/sh
                    - -c
                    - |
                      echo "Download pre-trained model into /workspace/model/qwen3.txt"
                      echo 'Qwen3-30b' > /workspace/model/qwen3.txt
                  volumeMounts:
                    - mountPath: /workspace/model
                      name: initializer
    - name: node
      dependsOn:
        - name: initializer
          status: Complete
      template:
        spec:
          parallelism: 3
          completions: 3
          template:
            spec:
              containers:
                - name: node
                  image: busybox
                  command:
                    - /bin/sh
                    - -c
                    - |
                      echo "Read pre-trained model"
                      cat /workspace/model/qwen3.txt
                      echo "Write model checkpoint to /workspace/checkpoints/model_node_${JOB_COMPLETION_INDEX}"
                      echo "Checkpoint: model_${JOB_COMPLETION_INDEX}" > /workspace/checkpoints/model_${JOB_COMPLETION_INDEX}.txt
                  volumeMounts:
                    - mountPath: /workspace/model
                      name: initializer
                    - mountPath: /workspace/checkpoints
                      name: checkpoints
```

TODO items:

- [x] Add unit tests for VolumeClaimPolicies controller changes
- [x] Add integration tests for VolumeClaimPolicies API
- [x] Add E2E tests for VolumeClaimPolicies API

/hold for review

#### Which issue(s) this PR fixes:

Fixes: https://github.com/kubernetes-sigs/jobset/issues/572

#### Special notes for your reviewer:

/cc @kannon92 @GiuseppeTT @ahg-g @imreddy13 @carreter @tenzen-y @astefanutti @Electronic-Waste @vsoch @xigang @akshaychitneni @shravan-achar @bigsur0 

#### Does this PR introduce a user-facing change?

```release-note
Add the VolumeClaimPolicies API to support stateful JobSet.
```

### CODE REVIEW

File: pkg/controllers/jobset_controller.go [hunk 1]
```diff
@@ -203,6 +204,14 @@ func (r *JobSetReconciler) reconcile(ctx context.Context, js *jobset.JobSet, upd
 		return ctrl.Result{}, err
 	}
 
+	// Reconcile VolumeClaimPolicies if it is set.
+	if len(js.Spec.VolumeClaimPolicies) > 0 {
```
Comment (kannon92) [inline pkg/controllers/jobset_controller.go:208]: Does it make sense to reconcile volume claim polcies when Jobset is suspended?

Comment (kannon92) [reply pkg/controllers/jobset_controller.go:208]: I guess they aren't any pods so they wouldn't bind but something to think through.

Comment (andreyvelich) [reply pkg/controllers/jobset_controller.go:208]: Hm, good point, I think we should create PVCs even if JobSet is created with suspended status, since this API object only consumes etcd memory. It looks like LWS controller doesn't check suspended state (e.g. Replicas=0) before creating PVCs: https://github.com/kubernetes-sigs/lws/blob/main/pkg/controllers/leaderworkerset_controller.go#L778-L780

Is my understand correct @ahg-g @Edwinhr716 @andyzhangx ?

Maybe @soltysh has some suggestions from Job controller perspective on this?

Comment (kannon92) [reply pkg/controllers/jobset_controller.go:208]: I think since we are only creating a global PVC for the entire workload this would be fine.

The main problem will be if we/when we start having PVC per ReplicatedJob. In that case, I think creating n replicas PVC could be a potential issue.

**Comment by kannon92** [comment]
Want to test out CoPilot reviews.

**Comment by kannon92** [comment]
I'm fine if you do this as a follow up but please don't forget to update our website with this feature.

**Comment by kannon92** [comment]
Copilot gets a F for this one.

**Comment by andreyvelich** [comment]
> I'm fine if you do this as a follow up but please don't forget to update our website with this feature.

Sure, let me track it in this issue: https://github.com/kubernetes-sigs/jobset/issues/1101

File: api/jobset/v1alpha2/jobset_types.go [hunk 2]
```diff
@@ -159,6 +160,16 @@ type JobSetSpec struct {
 	// +kubebuilder:validation:Minimum=0
 	// +optional
 	TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
+
+	// volumeClaimPolicies is a list of policies for persistent volume claims that pods are allowed
+	// to reference. JobSet controller automatically adds the required volume claims to the
+	// pod template. Every claim in this list must have at least one matching (by name)
+	// volumeMount in one container in the template.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
```
Comment (kannon92) [inline api/jobset/v1alpha2/jobset_types.go:168]: So I guess its good that this is immutable.

In some cases I'd like to be able to change JobSet objects while they are suspended. But in this case, I guess we are saying that this should never be modified even if the workload is suspended.

And the fact that we create even if suspended I guess this is fine.

Just calling this out.

Comment (andreyvelich) [reply api/jobset/v1alpha2/jobset_types.go:168]: Good point, I couldn't imagine a use-case when volumes might be mutable during workload suspension.

Theoretically, users might want to mount a new PVC after JobSet preemption, however that will require to make Pod spec `.volumes` mutable, which is impossible today, right?

File: pkg/controllers/volume_claim_policy.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+			}
+		} else {
+			log.V(2).Info("successfully created PVC", "pvc", klog.KObj(&pvc))
+		}
+	}
+
+	return nil
+}
+
+// generateSharedPVCName creates a name for the PVC in this format: <pvc-template-name>-<jobset-name>
```
Comment (kannon92) [inline pkg/controllers/volume_claim_policy.go]:
```suggestion
// generatePVCName creates a name for the PVC in this format: <pvc-template-name>-<jobset-name>
```

File: pkg/controllers/volume_claim_policy.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+		// Set PVC owner reference if retention policy is Delete.
+		if volumeClaimPolicy.RetentionPolicy != nil && volumeClaimPolicy.RetentionPolicy.WhenDeleted == jobset.RetentionPolicyDelete {
+			if err := ctrl.SetControllerReference(js, &pvc, r.Scheme); err != nil {
+				return err
+			}
+		}
+
+		// Create PVC if it doesn't exist.
+		// TODO (andreyvelich): Shall we consider to use workqueue.ParallelizeUntil, similar to rJobs creation?
```
Comment (kannon92) [inline pkg/controllers/volume_claim_policy.go]: Rather than TODO, could we create an issue to document follow ups?

File: test/integration/controller/jobset_controller_test.go [hunk 5]
```diff
@@ -1257,6 +1258,66 @@ var _ = ginkgo.Describe("JobSet controller", func() {
 				},
 			},
 		}),
+		ginkgo.Entry("jobset with VolumeClaimPolicies should create PVCs", &testCase{
```
Comment (kannon92) [inline test/integration/controller/jobset_controller_test.go]: We should test retention policies.

Comment (kannon92) [inline test/integration/controller/jobset_controller_test.go]: Let's have a test for suspended PVC also.

File: pkg/controllers/jobset_controller.go [hunk 6]
```diff
@@ -726,6 +735,11 @@ func constructJob(js *jobset.JobSet, rjob *jobset.ReplicatedJob, jobIdx int) *ba
 	jobsetSuspended := jobSetSuspended(js)
 	job.Spec.Suspend = ptr.To(jobsetSuspended)
 
+	// If VolumeClaimPolicies are set, update Job spec to set volumes and volumeMounts.
```
Comment (kannon92) [inline pkg/controllers/jobset_controller.go:738]: One issue we can have is if VolumeClaimTempaltes have syntax errors or not not fully create (create goes back with no error but other issues occur)

We would then add volumes to a job that actually does not have an existing volumeClaimTemplate.

File: api/jobset/v1alpha2/jobset_types.go (hunk 2 above)
Comment (kannon92) [inline api/jobset/v1alpha2/jobset_types.go:168]: Can you add a test case confirming immutability in integrations

File: api/jobset/v1alpha2/jobset_types.go [hunk 7]
```diff
@@ -458,6 +469,46 @@ type Coordinator struct {
...
+type VolumeClaimPolicy struct {
+	// templates is a list of shared PVC claims that ReplicatedJobs are allowed to reference.
+	// The JobSet controller is responsible for creating shared PVCs that can be mounted by
+	// multiple ReplicatedJobs. Every claim in this list must have a matching (by name)
+	// volumeMount in one container or initContainer in at least one ReplicatedJob template.
+	// ReplicatedJob template must not have volumes with the same name as defined in this template.
+	// Generated PVC naming convention: <claim-name>-<jobset-name>
+	// Example: "model-cache-trainjob" (shared volume across all ReplicatedJobs).
+	// +kubebuilder:validation:MaxItems=50
+	// +listType=atomic
```
Comment (kannon92) [inline api/jobset/v1alpha2/jobset_types.go:485]: I'm trying to fix the linter but I think it would complain here once I enabled Option or required settings.

Each field should have an optional tag.

File: api/jobset/v1alpha2/jobset_types.go [hunk 8]
```diff
@@ -458,6 +469,46 @@ type Coordinator struct {
...
+	// By default, all persistent volume claims are deleted once JobSet is deleted.
+	// +optional
+	RetentionPolicy *VolumeRetentionPolicy `json:"retentionPolicy,omitempty"`
+}
+
+// volumeRetentionPolicy defines the retention policy used for PVCs created from the JobSet VolumeClaimPolicies.
+type VolumeRetentionPolicy struct {
+	// whenDeleted specifies what happens to PVCs when JobSet is deleted.
+	// +kubebuilder:validation:Enum=Delete;Retain
+	// +kubebuilder:default=Delete
```
Comment (kannon92) [inline api/jobset/v1alpha2/jobset_types.go:498]: I think this should have +required

File: pkg/controllers/volume_claim_policy.go [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+		if err := r.Create(ctx, &pvc); err != nil {
+			if !apierrors.IsAlreadyExists(err) {
```
Comment (kannon92) [inline pkg/controllers/volume_claim_policy.go:62]: If retentionPolicy is Retain and someone creates a JobSet with a PVC.
And then they delete the JobSet with updated PVC templates will this cause any error?

File: pkg/webhooks/jobset_webhook.go [hunk 10]
```diff
@@ -409,6 +425,115 @@ func validateCoordinatorLabelValue(js *jobset.JobSet) error {
...
+	var allErrs []error
+	claimNames := sets.New[string]()
+
+	// Collect all claim names from templates.
+	for policyIdx, policy := range volumeClaimPolicies {
+		fieldPath := field.NewPath("spec", "volumeClaimPolicies").Index(policyIdx)
+		for templateIdx, template := range policy.Templates {
+			templateFieldPath := fieldPath.Child("template").Index(templateIdx)
+
+			// Validate template namespace.
```
Comment (kannon92) [inline pkg/webhooks/jobset_webhook.go]: Can we try and validate this with CEL on the API?

If not, we should also document at the API level the restrictions we put on VolumeClaimPolicies.

File: pkg/webhooks/jobset_webhook.go [hunk 11]
```diff
@@ -409,6 +425,115 @@ func validateCoordinatorLabelValue(js *jobset.JobSet) error {
...
+				))
+			}
+
+			// Validate DNS-1123 subdomain name
+			for _, err := range validation.IsDNS1123Subdomain(template.Name) {
+				allErrs = append(allErrs, field.Invalid(templateFieldPath.Child("name"), template.Name, err))
+			}
+
+			// Validate PVC name length limits
+			estimatedNameLength := len(template.Name) + len(js.Name) + 10 // Buffer for indices and separators
```
Comment (kannon92) [inline pkg/webhooks/jobset_webhook.go]: Not really obvious to me where 10 comes from.

What indices and separators are you talking about?

File: pkg/webhooks/jobset_webhook_test.go [hunk 12]
```diff
@@ -2032,11 +2101,453 @@ func TestValidateCreate(t *testing.T) {
...
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			want: errors.Join(fmt.Errorf("namespace cannot be set for the VolumeClaimPolicies template")),
```
Comment (kannon92) [inline pkg/webhooks/jobset_webhook_test.go]: Could we make the error message more descriptive on what template is failing (maybe name plus namespace?)

File: test/integration/controller/jobset_controller_test.go (hunk 5 above)
Comment (kannon92) [inline test/integration/controller/jobset_controller_test.go]: Can we add a test for retention polciy Retain but include a delete and recreate of a JobSet with update volume claims?

File: pkg/controllers/volume_claim_policy.go (hunk 9 above)
Comment (kannon92) [reply pkg/controllers/volume_claim_policy.go:62]: I guess this is a potential bug as PVC will be the old PVC.

File: test/e2e/e2e_test.go [hunk 13]
```diff
@@ -581,6 +582,38 @@ var _ = ginkgo.Describe("JobSet", func() {
...
+
+			// Create JobSet with VolumeClaimPolicies.
+			ginkgo.By("creating jobset with VolumeClaimPolicies")
+			js := volumeClaimTestJobSet(ns).Obj()
+
+			// Verify jobset created successfully.
+			ginkgo.By("checking that jobset creation succeeds")
+			gomega.Expect(k8sClient.Create(ctx, js)).Should(gomega.Succeed())
+
+			// Verify PVCs are created.
```
Comment (kannon92) [inline test/e2e/e2e_test.go:602]: There could be a potential flake issue here where JobSet does not created.

May be worth verifying that jobset is created before checking for PVC.

File: test/e2e/e2e_test.go [hunk 14]
```diff
@@ -581,6 +582,38 @@ var _ = ginkgo.Describe("JobSet", func() {
 			gomega.Expect(k8sClient.Delete(ctx, jobSet)).To(gomega.Succeed())
 		})
 	})
+	// This test runs JobSet with the VolumeClaimPolicies API.
+	ginkgo.When("VolumeClaimPolicies is enabled on JobSet", func() {
```
Comment (kannon92) [inline test/e2e/e2e_test.go:586]: Do you think adding a test case with multiple replicas would be useful??

Curious on how this work handles that all the jobs will write to the same volume claim.

**Review by kannon92 — COMMENTED** [review]
Really nice work.

I have a few comments on test cases and some potential edge cases.
But this is looking really good!

File: pkg/controllers/volume_claim_policy.go (hunk 4 above)
Comment (andreyvelich) [reply pkg/controllers/volume_claim_policy.go]: Sure, I am just unsure whether we want to parallelize PVC creation in the future.
@kannon92 Do you think that might be useful?

Comment (kannon92) [reply pkg/controllers/volume_claim_policy.go]: It feels a bit like YAGNI right now tbh.

File: pkg/controllers/jobset_controller.go (hunk 6 above)
Comment (andreyvelich) [reply pkg/controllers/jobset_controller.go:738]: I believe, for the syntax error we verify the DNS-1123 name: https://github.com/kubernetes-sigs/jobset/blob/07ced4f814f1cc66beb604d50df0f82293770e30/pkg/webhooks/jobset_webhook.go#L459

For other errors, we expose PVCCreationFailedReason to the JobSet events. Should that be sufficient?

File: pkg/controllers/volume_claim_policy.go (hunk 9 above)
Comment (andreyvelich) [reply pkg/controllers/volume_claim_policy.go:62]: How would they be able to modify the PVC template of running/suspended JobSet if API is immutable?

File: pkg/webhooks/jobset_webhook.go (hunk 11 above)
Comment (andreyvelich) [reply pkg/webhooks/jobset_webhook.go]: Maybe we should decrease it to 1?
Since we construct name as PVC template name + JobSet name.

File: pkg/webhooks/jobset_webhook_test.go (hunk 12 above)
Comment (andreyvelich) [reply pkg/webhooks/jobset_webhook_test.go]: This is the full error example:
```
            	Error:      	"spec.volumeClaimPolicies[0].template[0].namespace: Invalid value: \"invalid-namespace\": 
namespace cannot be set for the VolumeClaimPolicies template" does not contain "namespace cannot be set for
 the VolumeClaimPoliciesa template"
```

Since we bypass the index into `field.Invalid()`, the error will contain the index of the invalid slice.

@kannon92 Do you think that is not sufficient?

File: test/integration/controller/jobset_controller_test.go (hunk 5 above)
Comment (andreyvelich) [reply test/integration/controller/jobset_controller_test.go]: In this test do you want to modify the JobSet's pod volume mount path before JobSet is recreated?

File: test/e2e/e2e_test.go (hunk 14 above)
Comment (andreyvelich) [reply test/e2e/e2e_test.go:586]: Sure, let me create simple tests for multiple Replicas, maybe we can also verify Retain policy in E2Es.

File: pkg/controllers/volume_claim_policy.go (hunk 4 above)
Comment (andreyvelich) [reply pkg/controllers/volume_claim_policy.go]: Let me remove it for now, we can come back to it later.

File: pkg/webhooks/jobset_webhook_test.go (hunk 12 above)
Comment (kannon92) [reply pkg/webhooks/jobset_webhook_test.go]: Good point. This is probably fine.

File: test/integration/controller/jobset_controller_test.go (hunk 5 above)
Comment (kannon92) [reply test/integration/controller/jobset_controller_test.go]: I mean suspended JobSet

File: pkg/controllers/jobset_controller.go (hunk 6 above)
Comment (kannon92) [reply pkg/controllers/jobset_controller.go:738]: We still would addVolumes even if we got this error though?

The events will emit it but I don't think we check for success on adding Volumes.

File: api/jobset/v1alpha2/jobset_types.go (hunk 7 above)
Comment (kannon92) [reply api/jobset/v1alpha2/jobset_types.go:485]: Rebase should pick these up now.

File: pkg/controllers/volume_claim_policy.go (hunk 9 above)
Comment (kannon92) [reply pkg/controllers/volume_claim_policy.go:62]: In that case you are correct.

I am thinking Retain means that even if you remove the JobSet the PVC will still stay around.

So I'm thinking of the following scenario:

1) Create JobSet With PVC Retain.
2) Delete JobSet.
3) PVC should still be left around.
4) Create a new JobSet with same name with new PVC parameters.
5) Will the PVC get updated with new parameters or will it be left with parameters in 1?

File: pkg/webhooks/jobset_webhook.go (hunk 11 above)
Comment (kannon92) [reply pkg/webhooks/jobset_webhook.go]: could we just export the function that computes this and use that?

I don't get why we have to estimate this. We should know exactly what the length would be.

File: pkg/controllers/jobset_controller.go (hunk 6 above)
Comment (andreyvelich) [reply pkg/controllers/jobset_controller.go:738]: Yes, we still try to add Volumes to the Pod spec, and Pods should stuck in the Pending phase.
But that should be fine, what do you think?

File: pkg/controllers/volume_claim_policy.go (hunk 9 above)
Comment (andreyvelich) [reply pkg/controllers/volume_claim_policy.go:62]: oh, that is good point.
It looks like StatefulSet controller also doesn't validate it: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_pod_control.go#L354

I am wondering if we should introduce any additional validation for such use-cases?

File: pkg/controllers/jobset_controller.go (hunk 6 above)
Comment (kannon92) [reply pkg/controllers/jobset_controller.go:738]: I'm not sure they would be stuck in pending. Pending means that the scheduler wasn't able to find a node. I think it would go to ContainerCreateFailed or PodReadyToStartContainers=False.


I think they would end up in a failed state since the PVC wasn't created.

File: pkg/controllers/volume_claim_policy.go (hunk 9 above)
Comment (kannon92) [reply pkg/controllers/volume_claim_policy.go:62]: Yea maybe we can check that PVC matches at least.

File: pkg/webhooks/jobset_webhook.go (hunk 10 above)
Comment (andreyvelich) [reply pkg/webhooks/jobset_webhook.go]: Yes, CEL looks to be working fine for namespace.
Verifying the name uniqueness is not possible in CEL, since the complexity is O(n^2) which won't work in CEL.

File: pkg/controllers/volume_claim_policy.go (hunk 9 above)
Comment (andreyvelich) [reply pkg/controllers/volume_claim_policy.go:62]: @kannon92 Added validation of existing PVCs here: https://github.com/kubernetes-sigs/jobset/pull/1098/commits/500b7ab7338be4286387fa184458ca9517df4df9

**Comment by kannon92** [comment]
I'll review this on Friday.

File: pkg/controllers/jobset_controller.go (hunk 6 above)
Comment (andreyvelich) [reply pkg/controllers/jobset_controller.go:738]: Just verified it, and I can see that Pods will be in the Pending phase::
```yaml
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2025-12-29T19:10:26Z"
    message: '0/1 nodes are available: persistentvolumeclaim "test-volume" not
      found. preemption: 0/1 nodes are available: 1 Preemption is not helpful for
      scheduling.'
    reason: Unschedulable
    status: "False"
    type: PodScheduled
  phase: Pending
  qosClass: BestEffort
```

**Comment by andreyvelich** [comment]
@kannon92 Thanks for the review! I believe, I addressed all the suggestions, appreciate if you could take a look again!

File: api/jobset/v1alpha2/jobset_types.go [hunk 15]
```diff
@@ -159,6 +160,16 @@ type JobSetSpec struct {
...
 	TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
+
+	// volumeClaimPolicies is a list of policies for persistent volume claims that pods are allowed
+	// to reference. JobSet controller automatically adds the required volume claims to the
+	// pod template. Every claim in this list must have at least one matching (by name)
+	// volumeMount in one container in the template.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	// +optional
+	// +listType=atomic
+	// +kubebuilder:validation:MaxItems=50
```
Comment (kannon92) [inline api/jobset/v1alpha2/jobset_types.go:171]: can we have a integratoin test for webhook for this validation also?

File: pkg/controllers/jobset_controller.go (hunk 6 above)
Comment (kannon92) [reply pkg/controllers/jobset_controller.go:738]: sounds good then.

File: pkg/controllers/jobset_controller.go [hunk 16]
```diff
@@ -100,6 +100,7 @@ func NewJobSetReconciler(client client.Client, scheme *runtime.Scheme, record re
 //+kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=batch,resources=jobs/status,verbs=get;patch;update
 //+kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core,resources=persistentvolumeclaims,verbs=get;list;watch;create;update;patch;delete
```
Comment (kannon92) [inline pkg/controllers/jobset_controller.go]: Are all these rbac roles really required?

File: test/integration/controller/jobset_controller_test.go (hunk 5 above)
Comment (kannon92) [reply test/integration/controller/jobset_controller_test.go]: This test seems to be testing both retain and Deleted in the same test case. Could you update the title or consider separating these as two separate tests.

**Review by kannon92 — APPROVED** [review]
/lgtm

I have a few nits but I think this is good to go.

File: pkg/controllers/volume_claim_policy.go (hunk 9 above)
Comment (kannon92) [reply pkg/controllers/volume_claim_policy.go:62]: Looks good!

File: test/e2e/e2e_test.go (hunk 14 above)
Comment (kannon92) [reply test/e2e/e2e_test.go:586]: I missed this when I was reviewing. Was this done?

Comment (andreyvelich) [reply test/e2e/e2e_test.go:586]: Yes, I used 2 Replicas for the E2E: https://github.com/kubernetes-sigs/jobset/blob/c2d28e68563afe253385c0c2a20739fcc9ef22a0/test/e2e/e2e_test.go#L895

File: pkg/controllers/jobset_controller.go (hunk 16 above)
Comment (andreyvelich) [reply pkg/controllers/jobset_controller.go]: Good point, actually we probably need only `create`.

**Comment by andreyvelich** [comment]
> Should we update this function to include PersistentVolumeClaims?
> 
> https://github.com/kubernetes-sigs/jobset/blob/main/pkg/controllers/jobset_controller.go#L227-L235

Not really, since client uses cache for [this call](https://github.com/kubernetes-sigs/jobset/blob/133a94c457f7992616177f315d46e1023d836b7a/pkg/webhooks/jobset_webhook.go#L472), we just need to add `list` and `watch` RBAC.

Also, we don't trigger JobSet reconciler for Service or PVCs updates, so we might want to watch only for Jobs:
WDYT @kannon92 ?

```go
// SetupWithManager sets up the controller with the Manager.
func (r *JobSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
	return ctrl.NewControllerManagedBy(mgr).
		For(&jobset.JobSet{}).
		Owns(&batchv1.Job{}).
		Complete(r)
}
```

**Comment by andreyvelich** [comment]
@kannon92 Please let me know if you have more suggestion to this PR!

File: test/integration/webhook/jobset_webhook_test.go [hunk 17]
```diff
@@ -555,5 +555,203 @@ var _ = ginkgo.Describe("jobset webhook defaulting", func() {
...
+										},
+									},
+								},
+							}).
+							CompletionMode(batchv1.IndexedCompletion).Obj()).
+						Obj())
+			},
+			jobSetCreationShouldFail: true,
+		}),
+		ginkgo.Entry("VolumeClaimPolicies exceeding MaxItems=50 should be rejected", &testCase{
```
Comment (kannon92) [inline test/integration/webhook/jobset_webhook_test.go:676]: I'm having a hard time telling the difference between these two test cases.

Comment (kannon92) [reply test/integration/webhook/jobset_webhook_test.go:676]: Ah okay. I read the comments.

this seems fine.

**Comment by andreyvelich** [comment]
Thanks for your review and help @kannon92!

/hold cancel
